File: code/python/826.py
#https://leetcode.com/problems/most-profit-assigning-work/
class Solution:
    def maxProfitAssignment(self, difficulty: List[int], profit: List[int], worker: List[int]) -> int:
        res = idx = 0
        n = len(profit)
        diff_prof = list(zip(difficulty, profit))
        pq = []
        diff_prof.sort()
        worker.sort()
        for i in range(len(worker)):
            while idx < n and diff_prof[idx][0] <= worker[i]:
                heappush(pq, -diff_prof[idx][1])
                idx += 1
            if pq:
                res -= pq[0]
        return res

File: code/cpp/826.cpp
//https://leetcode.com/problems/most-profit-assigning-work/
class Solution {
public:
    int maxProfitAssignment(vector<int>& difficulty, vector<int>& profit, vector<int>& worker) {
        size_t res = 0, n = profit.size(), idx = 0;
        vector < pair<int, int>> diff_prof (n);
        priority_queue<int> pq;
        for (size_t i = 0; i < n; i++) {
            diff_prof[i] = { difficulty[i],profit[i] };
        }
        sort(diff_prof.begin(), diff_prof.end());
        sort(worker.begin(), worker.end());
        for (size_t i = 0; i < worker.size(); i++) {
            while (idx < n and diff_prof[idx].first <= worker[i]) {
                pq.push(diff_prof[idx].second);
                idx++;
            }
            if (!pq.empty()) res += pq.top();
        }
        return res;
    }
};
